js/modules/messages_data_migrator.js

```diff
@@ -264,19 +312,29 @@ const _dangerouslyFetchMessagesRequiringSchemaUpgradeWithoutIndex =
     const messagesStore = transaction.objectStore(MESSAGES_STORE_NAME);
 
     const excludeLowerBound = true;
-    const query = hasLastIndex
+    const range = hasLastIndex
       ? IDBKeyRange.lowerBound(lastIndex, excludeLowerBound)
       : undefined;
-    const request = messagesStore.getAll(query, count);
     return new Promise((resolve, reject) => {
-      request.onsuccess = event =>
-        resolve(event.target.result);
+      const items = [];
+      const request = messagesStore.openCursor(range);
+      request.onsuccess = (event) => {
+        const cursor = event.target.result;
+        const hasMoreData = Boolean(cursor);
+        if (!hasMoreData || items.length === count) {
+          resolve(items);
+          return;
+        }
+        const item = cursor.value;
+        items.push(item);
```

Comment:
`EventEmitter` style most certainly requires mutation like this. :0)

Reply:
Yep! I think it’s fine to have mutation internally in functions where required due to existing API (this case) or performance as long as the external interface returns immutable data 👍